core/src/main/java/com/linecorp/armeria/server/HttpResponseSubscriber.java

```diff
@@ -308,8 +324,8 @@ private void failAndRespond(Throwable cause, HttpStatus status, Http2Error error
 
         if (wroteNothing(state)) {
             // Did not write anything yet; we can send an error response instead of resetting the stream.
-            responseEncoder.writeHeaders(ctx, id, streamId, headers, false);
-            responseEncoder.writeData(ctx, id, streamId, content, true);
+            responseEncoder.writeHeaders(ctx, id, streamId, message.headers(), false);
+            responseEncoder.writeData(ctx, id, streamId, message.content(), true);
```

Comment:
Not related with this PR directly, but how about not writing the data frame if data is empty? i.e.

```java
final HttpHeaders headers = message.headers();
final HttpData content = message.content();
if (content.isEmpty()) {
    responseEncoder.writeHeaders(ctx, id, streamId, headers, true);
} else {
    responseEncoder.writeHeaders(ctx, id, streamId, headers, false);
    responseEncoder.writeData(ctx, id, streamId, content, true);
}
```